internal/pkg/agent/application/upgrade/upgrade_test.go

```diff
@@ -9,15 +9,93 @@ import (
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"runtime"
 	"strings"
 	"testing"
 
+	"github.com/gofrs/flock"
 	"github.com/stretchr/testify/require"
 
 	"github.com/elastic/elastic-agent/internal/pkg/release"
 	"github.com/elastic/elastic-agent/pkg/core/logger"
 )
 
+func Test_CopyFile(t *testing.T) {
+	l, _ := logger.New("test", false)
+	tt := []struct {
+		Name        string
+		From        string
+		To          string
+		IgnoreErr   bool
+		KeepOpen    bool
+		ExpectedErr bool
+	}{
+		{
+			"Existing, no onerr",
+			filepath.Join(".", "test", "case1", "README.md"),
+			filepath.Join(".", "test", "case1", "copy", "README.md"),
+			false,
+			false,
+			false,
+		},
+		{
+			"Existing but open",
+			filepath.Join(".", "test", "case2", "README.md"),
+			filepath.Join(".", "test", "case2", "copy", "README.md"),
+			false,
+			true,
+			runtime.GOOS == "windows", // this fails only on,
+		},
+		{
+			"Existing but open, ignore errors",
+			filepath.Join(".", "test", "case3", "README.md"),
+			filepath.Join(".", "test", "case3", "copy", "README.md"),
+			true,
+			true,
+			false,
+		},
+		{
+			"Not existing, accept errors",
+			filepath.Join(".", "test", "case4", "README.md"),
+			filepath.Join(".", "test", "case4", "copy", "README.md"),
+			false,
+			false,
+			true,
+		},
+		{
+			"Not existing, ignore errors",
+			filepath.Join(".", "test", "case4", "README.md"),
+			filepath.Join(".", "test", "case4", "copy", "README.md"),
+			true,
+			false,
+			false,
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.Name, func(t *testing.T) {
+			defer func() {
+				// cleanup
+				_ = os.RemoveAll(filepath.Dir(tc.To))
+			}()
+
+			var fl *flock.Flock
+			if tc.KeepOpen {
+				// this uses syscalls to create inter-process lock
+				fl = flock.New(tc.From)
+				_, err := fl.TryLock()
+				require.NoError(t, err)
+			}
+
+			err := copyDir(l, tc.From, tc.To, tc.IgnoreErr)
+			require.Equal(t, tc.ExpectedErr, err != nil)
+			if fl != nil {
+				require.NoError(t, fl.Unlock())
+			}
```

Comment:
Maybe invoke this via a `defer` so it will get executed even if the `require.Equal` on the previous line fails?